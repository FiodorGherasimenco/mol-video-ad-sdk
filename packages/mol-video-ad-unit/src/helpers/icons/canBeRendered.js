/* eslint-disable no-shadow */
const calculateArea = (element) => {
  const {
    height,
    width
  } = element.getBoundingClientRect();

  return height * width;
};

export const hasSpace = (newIcon, config) => {
  const {
    drawnIcons,
    placeholder
  } = config;
  const placeholderArea = calculateArea(placeholder);
  const iconArea = calculateArea(newIcon.element);
  const usedIconsArea = drawnIcons.reduce((accumulator, icon) => accumulator + icon.element, 0);

  return iconArea + usedIconsArea <= placeholderArea * 0.35;
};

export const withinBoundaries = (newIcon) => {
  const {
    left,
    top
  } = newIcon;

  return left >= 0 && top >= 0;
};

const right = ({left, width}) => left + width;
const left = ({left}) => left;
const top = ({top}) => top;
const bottom = ({top, height}) => top + height;
const overlap = (newIcon, drawnIcon) => left(newIcon) > right(drawnIcon) ||
  right(newIcon) < left(drawnIcon) ||
  bottom(newIcon) < top(drawnIcon) ||
  top(newIcon) > bottom(drawnIcon);

export const withoutOverlaps = (newIcon, {drawnIcons}) => !drawnIcons.some((drawnIcon) => overlap(newIcon, drawnIcon));

const canBeRendered = (newIcon, config) => hasSpace(newIcon, config) && withinBoundaries(newIcon) && withoutOverlaps(newIcon, config);

export default canBeRendered;
