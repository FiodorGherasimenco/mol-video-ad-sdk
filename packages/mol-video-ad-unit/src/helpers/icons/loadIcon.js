import createResource from '../resources/createResource';

// TODO: RENAME TO loadResource AND MOVE TO RESOURCES FOLDER
const loadIcon = (icon, {document, placeholder}) => new Promise((resolve, reject) => {
  try {
    const iconElement = createResource(document, icon);

    iconElement.addEventListener('error', () => {
      placeholder.removeChild(iconElement);
      iconElement.style.zIndex = 0;
      reject(new Error('Error loading icon'));
    });

    iconElement.addEventListener('load', () => {
      placeholder.removeChild(iconElement);
      iconElement.style.zIndex = 0;
      resolve(iconElement);
    });

    // Some browsers will not load the resource if they are not added to the DOM
    iconElement.style.zIndex = -9999;
    placeholder.appendChild(iconElement);
  } catch (error) {
    reject(error);
  }
});

export default loadIcon;
