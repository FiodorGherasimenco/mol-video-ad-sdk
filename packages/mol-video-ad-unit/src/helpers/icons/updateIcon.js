const isCustomXposition = (xPosition) => !['left', 'right'].includes(String(xPosition).toLowerCase());
const isCustomYPosition = (yPosition) => !['top', 'bottom'].includes(String(yPosition).toLowerCase());
const calculateIconLeft = (dynamicPos, iconWidth, drawnIcons, phWidth) => {
  const icons = drawnIcons.filter((icon) => icon.xPosition === dynamicPos);
  const drawnIconsWidth = icons.reduce((accumulator, icon) => accumulator + icon.width, 0);

  if (dynamicPos === 'left') {
    return drawnIconsWidth;
  }

  return phWidth - drawnIconsWidth - iconWidth;
};

const calculateIconTop = (dynamicPos, iconHeight, phHeight) => {
  if (dynamicPos === 'top') {
    return 0;
  }

  return phHeight - iconHeight;
};

const updateIcon = (icon, iconElement, {drawnIcons, placeholder}) => {
  const rect = iconElement.getBoundingClientRect();
  const phRect = placeholder.getBoundingClientRect();
  const width = icon.width || rect.width;
  const height = icon.height || rect.height;
  const xPosition = icon.xPosition || 'right';
  const yPosition = icon.yPosition || 'top';
  let left;
  let top;

  if (isCustomXposition(xPosition)) {
    left = xPosition;
  } else {
    left = calculateIconLeft(xPosition, width, drawnIcons, phRect.width);
  }

  if (isCustomYPosition(yPosition)) {
    top = yPosition;
  } else {
    top = calculateIconTop(yPosition, height, phRect.height);
  }

  return Object.assign(icon, {
    height,
    left,
    top,
    width
  });
};

export default updateIcon;
