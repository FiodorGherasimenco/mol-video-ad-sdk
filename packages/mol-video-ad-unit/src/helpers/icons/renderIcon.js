import loadResource from '../resources/loadResource';
import updateIcon from './updateIcon';
import canBeRendered from './canBeRendered';

const wrapWithClickThrough = (iconElement, icon, {onIconClick}) => {
  const anchor = document.createElement('A');

  // TODO: Test iconClickthrough logic
  if (icon.iconClickthrough) {
    anchor.href = icon.iconClickthrough;
    anchor.target = '_blank';
  }

  anchor.onclick = (event) => {
    if (Event.prototype.stopPropagation !== undefined) {
      event.stopPropagation();
    }

    onIconClick(icon);
  };

  anchor.appendChild(iconElement);

  return anchor;
};

const createIcon = async (icon, config) => {
  if (!icon.element) {
    const iconElement = await loadResource(icon, config);

    iconElement.height = '100%';
    iconElement.width = '100%';

    icon.element = wrapWithClickThrough(iconElement, icon, config);
  }

  return icon.element;
};

const updateIconElement = (iconElement, icon) => {
  const {
    height,
    width,
    left,
    top
  } = icon;

  iconElement.height = height;
  iconElement.width = width;
  iconElement.style.position = 'absolute';
  iconElement.style.left = `${left}px`;
  iconElement.style.top = `${top}px`;
  iconElement.style.height = `${height}px`;
  iconElement.style.width = `${width}px`;

  return iconElement;
};

const renderIcon = async (icon, config) => {
  const {placeholder} = config;
  const iconElement = await createIcon(icon, config);
  const updatedIcon = updateIcon(icon, iconElement, config);

  if (canBeRendered(updatedIcon, config)) {
    placeholder.appendChild(updateIconElement(iconElement, updatedIcon));
  } else {
    throw new Error('Icon can\'t be rendered');
  }

  return updatedIcon;
};

export default renderIcon;
