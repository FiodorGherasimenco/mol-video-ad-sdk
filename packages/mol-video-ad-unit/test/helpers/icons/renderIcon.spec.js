import renderIcon from '../../../src/helpers/icons/renderIcon';
import loadIcon from '../../../src/helpers/icons/loadIcon';
import updateIcon from '../../../src/helpers/icons/updateIcon';
import canBeRendered from '../../../src/helpers/icons/canBeRendered';

jest.mock('../../../src/helpers/icons/loadIcon');
jest.mock('../../../src/helpers/icons/updateIcon');
jest.mock('../../../src/helpers/icons/canBeRendered');

let config;
let icon;
let iconElement;
let placeholder;

beforeEach(() => {
  placeholder = document.createElement('DIV');
  iconElement = document.createElement('IMG');
  config = {
    placeholder
  };
  icon = {
    height: 5,
    left: 0,
    top: 0,
    width: 5,
    xPosition: 'left',
    yPosition: 'top'
  };
});

afterEach(() => {
  config = null;
  placeholder = null;
  iconElement = null;
  placeholder = null;
});

test('renderIcon must fail if there was a problem creating the icon', () => {
  const loadingError = new Error('problem loading icon');

  loadIcon.mockImplementation(() => Promise.reject(loadingError));
  expect(renderIcon(icon, config)).rejects.toBe(loadingError);
});

test('renderIcon must fail if the icon can not be rendered', () => {
  loadIcon.mockImplementation(() => Promise.resolve(iconElement));
  updateIcon.mockImplementation(() => icon);
  canBeRendered.mockImplementation(() => false);
  expect(renderIcon(icon, config)).rejects.toThrowErrorMatchingSnapshot();
});

test('must append the icon to the placeholder if three is no problem', async () => {
  loadIcon.mockImplementation(() => Promise.resolve(iconElement));
  updateIcon.mockImplementation(() => icon);
  canBeRendered.mockImplementation(() => true);

  await renderIcon(icon, config);

  expect(placeholder.contains(iconElement)).toBe(true);
});

test('renderIcon must reuse previously created icons', async () => {
  loadIcon.mockImplementation(() => Promise.resolve(iconElement));
  updateIcon.mockImplementation(() => icon);
  canBeRendered.mockImplementation(() => true);

  const renderedIcon = await renderIcon(icon, config);

  loadIcon.mockClear();

  await renderIcon(renderedIcon, config);

  expect(loadIcon).not.toHaveBeenCalled();

  expect(placeholder.contains(iconElement)).toBe(true);
});

test('renderIcon must return the updated icon', () => {
  const updatedIcon = Object.assign({}, icon);

  loadIcon.mockImplementation(() => Promise.resolve(iconElement));
  updateIcon.mockImplementation(() => updatedIcon);
  canBeRendered.mockImplementation(() => true);

  expect(renderIcon(icon, config)).resolves.toBe(updatedIcon);
});

test('renderIcon must style the icon Element', async () => {
  const updatedIcon = {
    height: 3,
    left: 1,
    top: 4,
    width: 6
  };

  loadIcon.mockImplementation(() => Promise.resolve(iconElement));
  updateIcon.mockImplementation(() => updatedIcon);
  canBeRendered.mockImplementation(() => true);

  await renderIcon(icon, config);

  expect(iconElement.height).toEqual(updatedIcon.height);
  expect(iconElement.width).toEqual(updatedIcon.width);
  expect(iconElement.style.position).toEqual('absolute');
  expect(iconElement.style.left).toEqual(`${updatedIcon.left}px`);
  expect(iconElement.style.top).toEqual(`${updatedIcon.top}px`);
  expect(iconElement.style.height).toEqual(`${updatedIcon.height}px`);
  expect(iconElement.style.width).toEqual(`${updatedIcon.width}px`);
});
