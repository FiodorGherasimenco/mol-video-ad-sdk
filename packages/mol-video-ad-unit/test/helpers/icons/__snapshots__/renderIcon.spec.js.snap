// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderIcon must fail if the icon can not be rendered 1`] = `"received is not a function"`;
